fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java

```diff
@@ -0,0 +1,340 @@
+import com.starrocks.catalog.ScalarType;
+import com.starrocks.catalog.Type;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.base.ColumnRefFactory;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.base.LogicalProperty;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CastOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Map;
+
+// For aggregate functions like `sum(expr + const)`, we can rewrite it to `sum(expr) + const * count()`
+// to reduce the overhead of a project calculation.
+public class RewriteSumByAssociativeRule extends TransformationRule {
+    public RewriteSumByAssociativeRule() {
+        super(RuleType.TF_REWRITE_SUM_BY_ASSOCIATIVE_RULE,
+                Pattern.create(OperatorType.LOGICAL_PROJECT)
+                        .addChildren(Pattern.create(OperatorType.LOGICAL_AGGR)
+                                .addChildren(Pattern.create(OperatorType.LOGICAL_PROJECT))));
+    }
+
+    @Override
+    public boolean check(final OptExpression input, OptimizerContext context) {
+        return !context.getSessionVariable().isDisableRewriteSumByAssociativeRule();
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        // let's use sum(id_int +1) as an example, suppose id_int is INT
+        /*
+         * Before
+         *
+         * LogicalProjectOperator {projection=[28: sum]}
+         * ->  LogicalAggregation {type=GLOBAL ,aggregations={28: sum=sum(27: expr)} ,groupKeys=[]}
+         *   ->  LogicalProjectOperator {projection=[add(cast(1: id_int as bigint(20)), 1)]}
+         *
+         * After
+         *
+         * LogicalProjectOperator {projection=[add(30: sum, multiply(31: count, 1))]
+         * ->  LogicalAggregation {type=GLOBAL ,aggregations={30: sum=sum(29: id_int), 31: count=count()} ,groupKeys=[]}
+         *   ->  LogicalProjectOperator {projection=[add(cast(1: id_int as bigint(20)), 1), 1: id_int]}
+         */
+        LogicalProjectOperator postAggProjectOperator = (LogicalProjectOperator) input.getOp();
```

Comment:
if there is no project operator resides above agg operator, can this optimization be applied?
so in such a situation, a trivial project operation always is interpolated above agg operator？

Reply:
> if there is no project operator resides above agg operator, can this optimization be applied? so in such a situation, a trivial project operation always is interpolated above agg operator？

I saw that there must be a project above agg in the logical plan we obtained initially(this project may be removed in the subsequent optimization), so I think it should be no problem to use this pattern matching here.